crates/bevy_ecs/src/query/iter.rs

```diff
@@ -157,6 +159,640 @@ impl<'w, 's, D: QueryData, F: QueryFilter> QueryIter<'w, 's, D, F> {
+
+        // SAFETY:
+        // `self.world` has permission to access the required components.
+        // The original query iter has not been iterated on, so no items are aliased from it.
+        let query_lens = unsafe {
+            query_lens_state.iter_unchecked_manual(
+                world,
+                world.last_change_tick(),
+                world.change_tick(),
+            )
+        };
+        let mut keyed_query: Vec<_> = query_lens
+            .map(|(key, entity)| (key, NeutralOrd(entity)))
+            .collect();
+        keyed_query.sort();
+        let entity_iter = keyed_query.into_iter().map(|(.., entity)| entity.0);
+        // SAFETY:
+        // `self.world` has permission to access the required components.
+        // Each lens query item is dropped before the respective actual query item is accessed.
+        unsafe {
+            QuerySortedIter::new(
+                world,
+                self.query_state,
+                entity_iter,
+                world.last_change_tick(),
+                world.change_tick(),
+            )
+        }
+    }
+
+    /// Sorts all query items into a new iterator, using the query lens as a key.
+    ///
+    /// This sort is unstable (i.e., may reorder equal elements).
+    ///
+    /// This uses [`slice::sort_unstable`] internally.
+    ///
+    /// Defining the lens works like [`transmute_lens`](crate::system::Query::transmute_lens).
+    /// This includes the allowed parameter type changes listed under [allowed transmutes]..
+    /// However, the lens uses the filter of the original query when present.
+    ///
+    /// [allowed transmutes]: crate::system::Query#allowed-transmutes
+    ///
+    /// # Panics
+    ///
+    /// This will panic if `next` has been called on `QueryIter` before, unless the underlying `Query` is empty.
+    ///
+    /// # Example
+    /// ```
+    /// # use bevy_ecs::prelude::*;
+    /// #
+    /// # let mut world = World::new();
+    /// #
+    /// # #[derive(Component)]
+    /// # struct PartMarker;
+    /// #
+    /// #[derive(Component, PartialEq, Eq, PartialOrd, Ord)]
+    /// enum Flying {
+    ///     Enabled,
+    ///     Disabled
+    /// };
+    ///
+    /// // We perform an unstable sort by a Component with few values.
+    /// fn system_1(query: Query<&Flying, With<PartMarker>>) {
+    ///     let part_values: Vec<&Flying> = query.iter().sort_unstable::<&Flying>().collect();
+    /// }
+    /// #
+    /// # let mut schedule = Schedule::default();
+    /// # schedule.add_systems((system_1));
+    /// # schedule.run(&mut world);
+    /// ```
+    pub fn sort_unstable<L: QueryData + 'w>(
+        self,
+    ) -> QuerySortedIter<
+        'w,
+        's,
+        D,
+        F,
+        impl ExactSizeIterator<Item = Entity> + DoubleEndedIterator + FusedIterator + 'w,
+    >
+    where
+        L::Item<'w>: Ord,
+    {
+        // On the first successful iteration of `QueryIterationCursor`, `archetype_entities` or `table_entities`
```

Comment:
just to be clear, if `next()` is called before `sort()` then this method is not valid because we could have duplicate mutable access to a component? (once in `next()` and once in the query_lens)

Or is it that we would be sorting the entire iterator (in the query_lens) even though part of the iterator has already been 'consumed'?

Reply:
You're correct, mutable reference aliasing would be possible. This is in addition of us having no immediate way of tracking how many iterations have occurred, without regressing the performance of more general iteration/sorted iteration.